akka-persistence/src/test/scala/akka/persistence/PersistentFSMActorSpec.scala

```diff
@@ -0,0 +1,339 @@
+ */
+class AbstractPersistentFSMSpec extends PersistentFSMActorSpec {
+  override def createPersistentFSMActor(persistenceId: String, reportActor: ActorRef): ActorRef =
+    system.actorOf(Props(new SampleJavaPersistentFsm(persistenceId, reportActor)))
+}
+
+object PersistentFSMActorSpec {
+  //State name
+  sealed trait UserState extends FSMState
+  case object LookingAround extends UserState {
+    override def identifier: String = "Looking Around"
+  }
+  case object Shopping extends UserState {
+    override def identifier: String = "Shopping"
+  }
+  case object Inactive extends UserState {
+    override def identifier: String = "Inactive"
+  }
+  case object Paid extends UserState {
+    override def identifier: String = "Paid"
+  }
+
+  case class Item(id: String, name: String, price: Float)
+
+  //State data
+  sealed trait ShoppingCart {
+    def addItem(item: Item): ShoppingCart
+    def empty(): ShoppingCart
+  }
+  case object EmptyShoppingCart extends ShoppingCart {
+    def addItem(item: Item) = NonEmptyShoppingCart(item::Nil)
+    def empty() = this
+  }
+  case class NonEmptyShoppingCart(items: Seq[Item]) extends ShoppingCart {
+    def addItem(item: Item) = NonEmptyShoppingCart(items :+ item)
+    def empty() = EmptyShoppingCart
+  }
+
+  //Commands
+  sealed trait Command
+  case class AddItem(item: Item) extends Command
+  case object Buy extends Command
+  case object Leave extends Command
+  case object GetCurrentCart extends Command
+
+  //Domain events
+  sealed trait DomainEvent
+  case class ItemAdded(item: Item) extends DomainEvent
+  case object OrderExecuted extends DomainEvent
+  case object OrderDiscarded extends DomainEvent
+
+  //Side effects - report events to be sent to some "Report Actor"
+  sealed trait ReportEvent
+  case class PurchaseWasMaid(items: Seq[Item])
+  case object ShoppingCardDiscarded
+
+  class SamplePersistentFsm(_persistenceId: String, reportActor: ActorRef) extends Actor with PersistentFsmActor[UserState, ShoppingCart, DomainEvent] {
+    override def persistenceId = _persistenceId
+
+    startWith(LookingAround, EmptyShoppingCart)
+
+    when(LookingAround) {
+      case Event(AddItem(item), _)      ⇒ goto(Shopping) applying ItemAdded(item)  forMax(1 seconds)
+      case Event(GetCurrentCart, data)  ⇒ stay replying data       //used for assertion of FSM's state during test
+    }
+
+    when(Shopping) {
+      case Event(AddItem(item), _)      ⇒ stay applying ItemAdded(item) forMax(1 seconds)
+      case Event(Buy, _)                ⇒ goto(Paid) applying OrderExecuted andThen { case NonEmptyShoppingCart(items) => reportActor ! PurchaseWasMaid(items) }
+      case Event(Leave, _)              ⇒ stop applying OrderDiscarded andThen { case _ => reportActor ! ShoppingCardDiscarded }
+      case Event(GetCurrentCart, data)  ⇒ stay replying data       //used for assertion of FSM's state during test
```

Comment:
I specifically wanted to test forMax, as it's part of FSM.State DSL. GetCurrentCart was needed in order to check the state of the FSM in some tests.
I don't use GetCurrentCart in timeout tests, to not affect the timeout - not an elegant solution, I know. This would be easier to implement if the timeouts would behave the way I've described.